sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/s3/S3FileSystem.java

```diff
@@ -548,12 +554,15 @@ CompleteMultipartUploadResponse multipartCopy(
                         Math.min(objectSize - 1, bytePosition + uploadBufferSizeBytes - 1)))
```

Comment:
Not sure, this might be another issue for multipart copies. Threshold to use multipart copy vs atomic copy is 5GB, using the upload buffer size (defaults to 5MB or 64MB depending on the environment) doesn't seem to make sense, right? That generates far more parts than necessary. Thoughts?

Reply:
Btw, here's docs for respective AWS limits: https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
5MB is the lower limit, so things will certainly work. But performance might suffer

Reply:
updated to use the MAX_COPY_OBJECT_SIZE_BYTES instead of the upload buffer size

Reply:
I can't find any good recommendations regarding part sizes of multipart copy... my gut feeling is that using the max possible we're on the other far end.  But it should be fine.
Have you been able to run some tests against S3 @spatel11?

Reply:
Yep, my test job that was failing before got done a few minutes ago.

This particular job produced 75 files ranging from 3.6 - 5.5 gb (so it's exercising both copy branches).

All files were copied from the temp location successfully (as opposed to before when it would fail on the multipart copy).

Furthermore, the files seemed structurally ok (I crawled them using AWS Glue and ran some simple queries).

Reply:
Awesome, great to hear!
Thanks so much for your contribution @spatel11 .

Reply:
Thanks for helping me!

Reply:
Perhaps, it would be useful to make it possible to configure the size of a part with multipart copy with CLI options (if not yet, sorry, I didn't check).

Reply:
Yes, I had the same thought but didn't wanna expand the scope of this fix. It's not yet possible to do that using S3Options. But I think it makes sense to add that similar to the config of uploadBufferSize.